### PR TITLE
Add a configuration option to disable XMLRPC search

### PR DESCRIPTION
**Pull request by dstufft**
This enables us to enable/disable the XMLRPC search API.

* Config: ``WAREHOUSE_XMLRPC_SEARCH=(y|n|yes|no|t|true|f|false|1|0``
  * Defaults to ``yes``, which will have the search API in a functional, non deprecated, state.
* Metrics: ``warehouse.xmlrpc.search.deprecated``
  * Incremented for every request that gets the deprecation message.

I left the wording purposely vague, which will allow us to either turn it on/off as a load shedding mechanism, or as part of a deprecation path.

### CODE REVIEW

**Review by ewdurbin — APPROVED** [review]

File: tests/unit/legacy/api/xmlrpc/test_xmlrpc.py [hunk 1]
```diff
@@ -114,6 +114,23 @@ def view(context, request):
...
+            pyramid_request.registry,
+            "settings",
+            {"warehouse.xmlrpc.search.enabled": False},
+        )
+        with pytest.raises(xmlrpc.XMLRPCWrappedError) as exc:
+            xmlrpc.search(pyramid_request, {"name": "foo", "summary": ["one", "two"]})
+
+        assert exc.value.faultString == (
+            "RuntimeError: This API has been deprecated due to unmanageable load. "
+            "Please Use the Simple or JSON API instead."
```
Comment (hugovk) [inline tests/unit/legacy/api/xmlrpc/test_xmlrpc.py:128]: Little nit:
```suggestion
            "Please use the Simple or JSON API instead."
```

**Comment by ewdurbin** [comment]
There is some concern about how this will render for `pip search` users at console.

An alternative to a fault string (though not optimal) is to just return a single result that will render something like

```shell
$ pip search fancypants
DEPRECATED (0.0.0)                                - The PyPI search endpoint that powers pip search has been deprecated.
```

File: tests/unit/legacy/api/xmlrpc/test_xmlrpc.py [hunk 2]
```diff
@@ -114,6 +114,23 @@ def view(context, request):
...
+        monkeypatch.setattr(
+            pyramid_request.registry,
+            "settings",
+            {"warehouse.xmlrpc.search.enabled": False},
+        )
+        with pytest.raises(xmlrpc.XMLRPCWrappedError) as exc:
+            xmlrpc.search(pyramid_request, {"name": "foo", "summary": ["one", "two"]})
+
+        assert exc.value.faultString == (
+            "RuntimeError: This API has been deprecated due to unmanageable load. "
```
Comment (ewdurbin) [inline tests/unit/legacy/api/xmlrpc/test_xmlrpc.py:127]: hmmm, this message seems _perfect_ for the final shut off, but in the meantime I might. suggest

```RuntimeError: This API has been temporarily disabled to to unmanageable load and will be deprecated in near future. Please use the Simple or JSON API instead.```

**Comment by dstufft** [comment]
This is what pip does with the hard failure: 

```
❯ pip search --index http://localhost/pypi pip
ERROR: Exception:
Traceback (most recent call last):
  File "/Users/dstufft/.pyenv/versions/3.9.0/lib/python3.9/site-packages/pip/_internal/cli/base_command.py", line 228, in _main
    status = self.run(options, args)
  File "/Users/dstufft/.pyenv/versions/3.9.0/lib/python3.9/site-packages/pip/_internal/commands/search.py", line 60, in run
    pypi_hits = self.search(query, options)
  File "/Users/dstufft/.pyenv/versions/3.9.0/lib/python3.9/site-packages/pip/_internal/commands/search.py", line 80, in search
    hits = pypi.search({'name': query, 'summary': query}, 'or')
  File "/Users/dstufft/.pyenv/versions/3.9.0/lib/python3.9/xmlrpc/client.py", line 1116, in __call__
    return self.__send(self.__name, args)
  File "/Users/dstufft/.pyenv/versions/3.9.0/lib/python3.9/xmlrpc/client.py", line 1456, in __request
    response = self.__transport.request(
  File "/Users/dstufft/.pyenv/versions/3.9.0/lib/python3.9/site-packages/pip/_internal/network/xmlrpc.py", line 45, in request
    return self.parse_response(response.raw)
  File "/Users/dstufft/.pyenv/versions/3.9.0/lib/python3.9/xmlrpc/client.py", line 1348, in parse_response
    return u.close()
  File "/Users/dstufft/.pyenv/versions/3.9.0/lib/python3.9/xmlrpc/client.py", line 662, in close
    raise Fault(**self._stack[0])
xmlrpc.client.Fault: <Fault -32500: 'RuntimeError: This API has been deprecated due to unmanageable load. Please Use the Simple or JSON API instead.'>
```

That's not *great*, but my concern is that if we just return nonsense results, then we risk silently breaking people who are relying on this API in an automated fashion. I think it's a lot better to explicitly error than to just silently do the wrong thing.

**Comment by ewdurbin** [comment]
@dstufft just to confirm, pip issues a non-zero exit status in that case as well?

**Comment by dstufft** [comment]
```
❯ pip search --index http://localhost/pypi pip
ERROR: Exception:
Traceback (most recent call last):
  File "/Users/dstufft/.pyenv/versions/3.8.2/lib/python3.8/site-packages/pip/_internal/cli/base_command.py", line 188, in _main
    status = self.run(options, args)
  File "/Users/dstufft/.pyenv/versions/3.8.2/lib/python3.8/site-packages/pip/_internal/commands/search.py", line 52, in run
    pypi_hits = self.search(query, options)
  File "/Users/dstufft/.pyenv/versions/3.8.2/lib/python3.8/site-packages/pip/_internal/commands/search.py", line 71, in search
    hits = pypi.search({'name': query, 'summary': query}, 'or')
  File "/Users/dstufft/.pyenv/versions/3.8.2/lib/python3.8/xmlrpc/client.py", line 1109, in __call__
    return self.__send(self.__name, args)
  File "/Users/dstufft/.pyenv/versions/3.8.2/lib/python3.8/xmlrpc/client.py", line 1450, in __request
    response = self.__transport.request(
  File "/Users/dstufft/.pyenv/versions/3.8.2/lib/python3.8/site-packages/pip/_internal/network/xmlrpc.py", line 38, in request
    return self.parse_response(response.raw)
  File "/Users/dstufft/.pyenv/versions/3.8.2/lib/python3.8/xmlrpc/client.py", line 1341, in parse_response
    return u.close()
  File "/Users/dstufft/.pyenv/versions/3.8.2/lib/python3.8/xmlrpc/client.py", line 655, in close
    raise Fault(**self._stack[0])
xmlrpc.client.Fault: <Fault -32500: 'RuntimeError: This API has been temporarily disabled due to unmanageable load and will be deprecated in near future. Please use the Simple or JSON API instead.'>

❯ echo $?     
2
```

**Review by ewdurbin — APPROVED** [review]

**Pull request merged**